File: src/main/scala/com/abaddon83/cardsGames/burracoGames/queries/FindBurracoGameEndedQuery.scala
package com.abaddon83.cardsGames.burracoGames.queries

import com.abaddon83.cardsGames.burracoGames.domainModels.burracoGames.ended.BurracoGameEnded
import com.abaddon83.cardsGames.burracoGames.ports.BurracoGameRepositoryPort
import com.abaddon83.cardsGames.shares.games.GameIdentity
import com.abaddon83.libs.cqs.queries.{Query, QueryHandler}

import scala.concurrent.Future

case class FindBurracoGameEndedQuery(
                                        gameIdentity: GameIdentity
                                      ) extends Query {
}

class FindBurracoGameEndedHandler(
                                     gameRepositoryPort: BurracoGameRepositoryPort
                                   ) extends QueryHandler[FindBurracoGameEndedQuery, BurracoGameEnded] {
  override def handleAsync(query: FindBurracoGameEndedQuery): Future[BurracoGameEnded] = {
    gameRepositoryPort.findBurracoBurracoGameEndedBy(query.gameIdentity)
  }

  override def handle(query: FindBurracoGameEndedQuery): BurracoGameEnded = ???
}

File: src/main/kotlin/com/abaddon83/burracoGame/queries/FindBurracoGameEndedQuery.kt
package com.abaddon83.burracoGame.queries

import com.abaddon83.burracoGame.domainModels.burracoGameendeds.BurracoGameEnded
import com.abaddon83.burracoGame.ports.BurracoGameRepositoryPort
import com.abaddon83.burracoGame.shared.games.GameIdentity
import com.abaddon83.utils.cqs.queries.Query
import com.abaddon83.utils.cqs.queries.QueryHandler
import kotlinx.coroutines.runBlocking
import org.koin.core.KoinComponent
import org.koin.core.inject

data class FindBurracoGameEndedQuery(val gameIdentity: GameIdentity): Query<BurracoGameEnded?> {
}

class FindBurracoGameEndedHandler(): QueryHandler<FindBurracoGameEndedQuery, BurracoGameEnded?>, KoinComponent {

    private val repository: BurracoGameRepositoryPort by inject()

    override fun handle(query: FindBurracoGameEndedQuery): BurracoGameEnded? {
        return runBlocking { repository.findBurracoGameEndedBy(query.gameIdentity) }
    }

    override suspend fun handleAsync(query: FindBurracoGameEndedQuery): BurracoGameEnded? {
        return repository.findBurracoGameEndedBy(query.gameIdentity)
    }
}